### PR TITLE
Fix parallel.do with batch norm

### PR DESCRIPTION
**Pull request by reyoung**
Related issue #8153

### CODE REVIEW

File: paddle/operators/parallel_do_op.cc [hunk 1]
```diff
@@ -248,6 +248,8 @@ class ParallelDoGradOp : public framework::OperatorBase {
                       const std::vector<framework::Scope *> &sub_scopes,
                       const platform::PlaceList &places) const {
     for (auto &s : Outputs(framework::GradVarName(kParameters))) {
+      VLOG(10) << "Accumulating " << s;
+      if (s == framework::kEmptyVarName) continue;
```
Comment (reyoung) [inline paddle/operators/parallel_do_op.cc]: Do not accumulate @EMPTY@

File: paddle/operators/parallel_do_op.cc [hunk 2]
```diff
@@ -360,10 +355,14 @@ class ParallelDoGradOpShapeInference : public framework::InferShapeBase {
       ctx->SetDims({ig_name}, {i_dims[i]});
     }
 
-    if (ctx->HasInputs(kParameters)) {
-      PADDLE_ENFORCE(ctx->HasOutputs(framework::GradVarName(kParameters)));
-      ctx->SetOutputsDim(framework::GradVarName(kParameters),
-                         ctx->GetInputsDim(kParameters));
+    auto p_dims = ctx->GetInputsDim(kParameters);
```
Comment (reyoung) [inline paddle/operators/parallel_do_op.cc]: If parameter gradient is empty, do not infer shape.

File: python/paddle/v2/fluid/layers/control_flow.py [hunk 3]
```diff
@@ -274,21 +274,20 @@ def get_parameters(self):
         parent_block = self.parent_block()
 
         local_inputs = set()
-
```
Comment (reyoung) [inline python/paddle/v2/fluid/layers/control_flow.py]: The previous logic cannot calculate parameters that used and updated by the same operator.

**Comment by tonyyang-svail** [comment]
For the purpose of this PR, please ignore the nccl error: `unhandled cuda error`. I suspect the reason to be paddle compiling with nccl1.3 but linking with nccl2.

**Comment by reyoung** [comment]
Merged by #8249

**Pull request closed without merging**